### PR TITLE
fix: prayer onboarding logic

### PR DESCRIPTION
**Pull request by christinamegura**
## DESCRIPTION

This PR fixes a bug that caused onboarding to never show. 

### What does this PR do, or why is it needed?
1. This PR creates a separate async function to be called from `useEffect`.
2. The second argument passed in `useEffect` is the `onboardingKey` element. 
3. Async value is set after modal is opened.

## TODO:

- [ ] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] Closes [tag-issues-here]
- [ ] No new warnings in tests, in storybook, and in-app
- [ ] Upload GIF(s) of iOS and Android
- [ ] Set a relevant reviewer

## REVIEW:

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

File: packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js [hunk 1]
```diff
@@ -19,20 +19,40 @@ function PrayerFeatureConnected({
...
+//     })();
+//   }, []);
 
   const handleOpenTo = (index = 0) => {
     setSwiperIndex(index);
-    AsyncStorage.setItem(onboardingKey, 'YES');
+    // AsyncStorage.setItem(onboardingKey, 'YES');
     setModalOpened(true);
+    // can I set this after I call these functions? Will it cause a re-render?
+    AsyncStorage.setItem(onboardingKey, 'YES');
```
Comment (christinamegura) [inline packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js:55]: @vinnyjth can you take a look at how I'm approaching this? The problem boils down to when`setModalOpened` is called, `useEffect` re-renders with the new onboarding key thus onboarding is never seen

Comment (vinnyjth) [reply packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js:55]: Can you put the `onboardingKey` inside the array on line 40 that controls the watched props for `useEffect`?

File: packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js [hunk 2]
```diff
@@ -22,17 +22,19 @@ function PrayerFeatureConnected({
...
+    setShouldShowOnboarding(!hasSeenOnboarding);
+  })();
+
   useEffect(() => {
-    (async () => {
-      const hasSeenOnboarding = await AsyncStorage.getItem(onboardingKey);
-      setShouldShowOnboarding(!hasSeenOnboarding);
-    })();
-  });
+    getOnboardingKey();
```
Comment (conrad-vanl) [inline packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js:31]: What are you trying to achieve by defining a `getOnboradingKey` function and splitting this code out into it?

File: packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js [hunk 3]
```diff
@@ -22,17 +22,19 @@ function PrayerFeatureConnected({
...
+  }, [onboardingKey]);
```
Comment (conrad-vanl) [inline packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js:32]: What does adding `[onboardingKey]` change / do?

**Review by conrad-vanl — COMMENTED** [review]
@christinamegura thanks for getting this up! What's the current state of this PR? I noticed it's still "In Draft" stage, does that mean you're still working on it or is it ready for review?

File: packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js (hunk 2 above)
Comment (christinamegura) [reply packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js:31]: By defining `getOnboardingKey` I'm creating a separate async function to call from `useEffect`. Async functions return promises and `useEffect` doesn't expect the callback function to return a promise.

File: packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js (hunk 3 above)
Comment (christinamegura) [reply packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js:32]: Passing `onboardingKey` as the second argument means the effect will run on initial render and on any (state) change of `onboardingKey`. With no second arguments passed, `useEffect` would run on every state change. With an empty array passed as a second argument, `useEffect` will run on the initial mount.  We want `useEffect` run on initial mount and whenever there are changes to `onboardingKey`.

**Comment by christinamegura** [comment]
> @christinamegura thanks for getting this up! What's the current state of this PR? I noticed it's still "In Draft" stage, does that mean you're still working on it or is it ready for review?

Almost ready. I'm having a small issue with my sim and need to write a snippet on how to test in the PR description.

**Comment by vinnyjth** [comment]
Getting this error when booting the app

Screen Shot 2020-10-19 at 9 11 42 AM

File: packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js [hunk 4]
```diff
@@ -22,17 +22,19 @@ function PrayerFeatureConnected({
   const [shouldShowOnboarding, setShouldShowOnboarding] = useState(false);
   const onboardingKey = `${featureId}-seenOnboarding`;
 
+  const getOnboardingKey = (async () => {
```
Comment (vinnyjth) [inline packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js:25]: I believe there's a subtle bug in how this function is defined that's causing issues. I would look into what `getOnboardingKey` ends up being (what type of 'thing'), and compare that to what you are expecting :)

**Comment by vinnyjth** [comment]
@christinamegura Let me know when you want me to look at this again :)

**Comment by christinamegura** [comment]
@vinnyjth I can use your eyes on this again, please.

File: packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js [hunk 5]
```diff
@@ -22,17 +22,24 @@ function PrayerFeatureConnected({
...
-  useEffect(() => {
-    (async () => {
+  const getOnboardingKey = async () => {
+    try {
       const hasSeenOnboarding = await AsyncStorage.getItem(onboardingKey);
       setShouldShowOnboarding(!hasSeenOnboarding);
-    })();
-  });
+      } catch (e) {
+        console.error(e);
```
Comment (vinnyjth) [inline packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js:30]: Out of curiosity, was there an error you were encountering here that you were trying to catch? This isn't a bad move, just curious :)

File: packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js [hunk 6]
```diff
@@ -22,17 +22,24 @@ function PrayerFeatureConnected({
   const [shouldShowOnboarding, setShouldShowOnboarding] = useState(false);
   const onboardingKey = `${featureId}-seenOnboarding`;
 
-  useEffect(() => {
-    (async () => {
+  const getOnboardingKey = async () => {
```
Comment (vinnyjth) [inline packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js:25]: @christinamegura What are your thoughts on renaming this function? While it is technically `getting the onboarding key`, it seems like its more important intent is to set if onboarding has been seen or not.

**Comment by vinnyjth** [comment]
This is working well for me @christinamegura! I did notice that if I opened the prayer, got the onboarding flow, and then closed prayer, I would continue to get prayer onboarding until I closed my app, pulled to refetch, or anything that remounted my PrayerFeature.

I think it's okay if we push the solution to that issue to another PR, since I imagine most people won't experience the bug.

**Comment by vinnyjth** [comment]
I left one question about the naming of the `getOnboardingKey` function. Once we address that, + tests and linter, we'll be good to merge!

File: packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js (hunk 5 above)
Comment (christinamegura) [reply packages/apollos-ui-connected/src/PrayerListFeatureConnected/index.js:30]: I was not encountering an error but I thought I could improve this. A promise can be in a pending, fulfilled, or rejected state and I wasn't handling a rejected promise anywhere.

**Review by vinnyjth — APPROVED** [review]
